File: evaluator/009.session-expiry-testability/tests/test_session_manager.cc
#include "session_manager.h"

#include <cstdint>
#include <memory>
#include <string>

#include <gtest/gtest.h>

#include "time_source.h"

namespace {

class ManualTimeSource final : public nitr::case009::TimeSource {
 public:
  explicit ManualTimeSource(std::int64_t now_seconds)
      : now_seconds_(now_seconds) {}

  std::int64_t NowSeconds() const override {
    return now_seconds_;
  }

  void AdvanceSeconds(std::int64_t delta_seconds) {
    now_seconds_ += delta_seconds;
  }

 private:
  std::int64_t now_seconds_;
};

TEST(Case009SessionManager, CreatedSessionIsImmediatelyValid) {
  auto clock = std::make_shared<ManualTimeSource>(100);
  nitr::case009::SessionManager manager(10, clock);

  EXPECT_TRUE(manager.CreateSession("s1"));
  EXPECT_TRUE(manager.IsValid("s1"));
}

TEST(Case009SessionManager, ExpiresAtBoundary) {
  auto clock = std::make_shared<ManualTimeSource>(100);
  nitr::case009::SessionManager manager(10, clock);

  EXPECT_TRUE(manager.CreateSession("s1"));
  clock->AdvanceSeconds(9);
  EXPECT_TRUE(manager.IsValid("s1"));

  clock->AdvanceSeconds(1);
  EXPECT_FALSE(manager.IsValid("s1"));
}

TEST(Case009SessionManager, RefreshRestartsWindow) {
  auto clock = std::make_shared<ManualTimeSource>(100);
  nitr::case009::SessionManager manager(10, clock);

  EXPECT_TRUE(manager.CreateSession("s1"));
  clock->AdvanceSeconds(5);
  EXPECT_TRUE(manager.RefreshSession("s1"));

  clock->AdvanceSeconds(9);
  EXPECT_TRUE(manager.IsValid("s1"));

  clock->AdvanceSeconds(1);
  EXPECT_FALSE(manager.IsValid("s1"));
}

TEST(Case009SessionManager, CannotRefreshExpiredSession) {
  auto clock = std::make_shared<ManualTimeSource>(100);
  nitr::case009::SessionManager manager(10, clock);

  EXPECT_TRUE(manager.CreateSession("s1"));
  clock->AdvanceSeconds(10);
  EXPECT_FALSE(manager.RefreshSession("s1"));
}

TEST(Case009SessionManager, RemoveInvalidatesSession) {
  auto clock = std::make_shared<ManualTimeSource>(100);
  nitr::case009::SessionManager manager(10, clock);

  EXPECT_TRUE(manager.CreateSession("s1"));
  EXPECT_TRUE(manager.RemoveSession("s1"));
  EXPECT_FALSE(manager.IsValid("s1"));
}

}  // namespace

File: evaluator/030.session-expiry-testability-python/tests/test_session_manager.py
#!/usr/bin/env python3

from __future__ import annotations

import argparse
import sys
import unittest
from pathlib import Path

ARG_PARSER = argparse.ArgumentParser()
ARG_PARSER.add_argument("--case_root", required=True)
ARGS = ARG_PARSER.parse_args()

CASE_ROOT = Path(ARGS.case_root).resolve()
if str(CASE_ROOT) not in sys.path:
    sys.path.insert(0, str(CASE_ROOT))

from src.session_manager import SessionManager
from src.time_source import TimeSource


class ManualTimeSource(TimeSource):
    def __init__(self, now_seconds: int) -> None:
        self._now_seconds = now_seconds

    def now_seconds(self) -> int:
        return self._now_seconds

    def advance_seconds(self, delta_seconds: int) -> None:
        self._now_seconds += delta_seconds


class SessionManagerTests(unittest.TestCase):
    def test_created_session_is_immediately_valid(self) -> None:
        clock = ManualTimeSource(100)
        manager = SessionManager(10, time_source=clock)

        self.assertTrue(manager.create_session("s1"))
        self.assertTrue(manager.is_valid("s1"))

    def test_expires_at_boundary(self) -> None:
        clock = ManualTimeSource(100)
        manager = SessionManager(10, time_source=clock)

        self.assertTrue(manager.create_session("s1"))
        clock.advance_seconds(9)
        self.assertTrue(manager.is_valid("s1"))

        clock.advance_seconds(1)
        self.assertFalse(manager.is_valid("s1"))

    def test_refresh_restarts_window(self) -> None:
        clock = ManualTimeSource(100)
        manager = SessionManager(10, time_source=clock)

        self.assertTrue(manager.create_session("s1"))
        clock.advance_seconds(5)
        self.assertTrue(manager.refresh_session("s1"))

        clock.advance_seconds(9)
        self.assertTrue(manager.is_valid("s1"))

        clock.advance_seconds(1)
        self.assertFalse(manager.is_valid("s1"))

    def test_cannot_refresh_expired_session(self) -> None:
        clock = ManualTimeSource(100)
        manager = SessionManager(10, time_source=clock)

        self.assertTrue(manager.create_session("s1"))
        clock.advance_seconds(10)
        self.assertFalse(manager.refresh_session("s1"))

    def test_remove_invalidates_session(self) -> None:
        clock = ManualTimeSource(100)
        manager = SessionManager(10, time_source=clock)

        self.assertTrue(manager.create_session("s1"))
        self.assertTrue(manager.remove_session("s1"))
        self.assertFalse(manager.is_valid("s1"))


if __name__ == "__main__":
    unittest.main(argv=[sys.argv[0]])
